src/WebManager/app/src/components/multi-tag-list.tsx

```diff
@@ -0,0 +1,329 @@
+                checkedTags: tempArray
+            } as IMultiTagListState);
+
+        } else {
+            this.hash[e.target.value] = "";
+            let valueIndex = tempArray.indexOf(e.target.value);
+            tempArray.splice(valueIndex, 1);
+            this.setState({
+                checkedTags: tempArray
+            } as IMultiTagListState);
+        }
+    }
+
+    componentDidMount(): void {
+        this.getMoreTags();
+    }
+
+    componentWillUnmount(): void {
+        if (this.cancel) {
+            this.cancel.cancel("component unmounting");
+            this.cancel = null;
+        }
+    }
+
+    getMoreTags(): void {
+        if (this.cancel) {
+            return;
+        }
+        if (!this.state.hasMoreTags) {
+            return;
+        }
+
+        let last: string = null;
+        if (this.state.tags != null) {
+            last = this.state.tags[this.state.tags.length - 1];
+        }
+        this.cancel = this.props.service.createCancelToken();
+        this.props.service.getTagsForRepo(this.props.params.repositoryName, 10, last, this.cancel.token)
+            .then(value => {
+                this.cancel = null;
+                if (!value) return;
+
+                this.setState((prevState, props) => {
+                    if (prevState.tags == null) {
+                        prevState.tags = [];
+                    }
+                    for (let tag of value.tags) {
+                        prevState.tags.push(tag);
+                    }
+
+                    prevState.hasMoreTags = value.httpLink !== undefined;
+
+                    return prevState;
+                });
+            }).catch(err => {
+                this.cancel = null;
+                this.setState({
+                    error: err.toString()
+                } as IMultiTagListState);
+
+                if (this.props.onLoadFailure) {
+                    this.props.onLoadFailure(err);
+                }
+            });
+    }
+
+    getDigestAndSize(name: string): void {
+        this.cancel = this.props.service.createCancelToken();
+        this.props.service.getManifestHeaders(this.props.params.repositoryName, name, this.cancel.token)
+            .then(value => {
+                this.cancel = null;
+                if (!value) return
+
+                this.setState((prevState, props) => {
+                    if (prevState.multiManifestTags == null) {
+                        prevState.multiManifestTags = [];
+                    }
+                    var tet = this.hash[name];
+                    if (tet == undefined || tet == null || tet == "") {
+                        if (this.followsConvention(name)) {
+                            var aux = name.split("-");
+                            var arch = aux[2];
+                            var opS = aux[1];
+                        }
+                        else {
+                            alert("Please specify the architecture and os");
```

Comment:
Not necessarily bad, but alternatively consider changing all calls to `alert()` for error messages nearby the affected fields. Typically form validation can take care of this, and you can display something like text in red next to the input boxes indicating what the error is.